core/src/main/kotlin/io/github/kkafka/producer/transaction.kt

```diff
@@ -0,0 +1,110 @@
+
+        // Get the offsets of the partitions so we can commit them as part of the transaction
+        val offsets = consumerTopicPartitions.associateWith { OffsetAndMetadata(consumer.position(it)) }
+
+        val groupId = consumer.groupMetadata().groupId()
+
+        // Make sure that we got all the needed partitions
+        if (offsets.size != consumerTopics.size) {
+            throw TransactionAbortedException(
+                "consumer in transaction is not properly subscribed to all topics given for transaction"
+            )
+        }
+
+        // Commit the consumer's offsets
+        sendOffsetsToTransaction(offsets, groupId)
+
+        // Successful transaction
+        commitTransaction()
+
+        res
+    } catch (e: Exception) {
+        throw handleTransactionException(e)
+    }
+}
+
+/**
+ * Consume and produce transactionally on a single topic.
+ * See [transaction].
+ */
+public fun <K, V, T> KafkaProducer<K, V>.transaction(
+    consumer: KafkaConsumer<*, *>,
+    consumerTopic: String,
+    block: KafkaProducer<K, V>.() -> T,
+): T = transaction(consumer, setOf(consumerTopic), block)
+
+/**
+ * Produce transactionally with an arbitrary block of code.
+ * If an exception occurs during the execution of [block] or the transaction commit, the transaction is aborted,
+ * otherwise it is committed and the result of [block] is returned.
+ *
+ * If any of [ProducerFencedException], [OutOfOrderSequenceException], [UnsupportedVersionException] or
+ * [AuthorizationException] is thrown, the producer is in an invalid state and is automatically closed.
+ * For any other exception, the transaction is aborted and may be tried again.
+ */
+public fun <K, V, T> KafkaProducer<K, V>.transaction(block: KafkaProducer<K, V>.() -> T): T {
+    beginTransaction()
+    return try {
+        val res = block(this)
+        commitTransaction()
+        res
+    } catch (e: Exception) {
+        throw handleTransactionException(e)
+    }
+}
+
+/*
+ * If any of [ProducerFencedException], [OutOfOrderSequenceException], [UnsupportedVersionException] or
+ * [AuthorizationException] is thrown, the producer is in an invalid state and is automatically closed.
+ * For any other exception, the transaction is aborted and may be tried again.
+ */
+private fun <K, V> KafkaProducer<K, V>.handleTransactionException(e: Exception): KafkaException {
+    return when (e) {
+        is ProducerFencedException, is OutOfOrderSequenceException,
+        is AuthorizationException, is UnsupportedVersionException -> {
+            close()
+            ProducerClosedException("A fatal exception occurred during transaction and the producer had to be closed.")
+        }
+        else -> {
+            abortTransaction()
+            TransactionAbortedException("An exception occurred during transaction. The transaction may be tried again.")
+        }
+    }
+}
```

Comment:
- are you sure we want to catch these and rethrow a different exception? The user might wrap _our_ code in a `try {} catch (e: OutOfOrderSequenceException)` or `AuthorizationException`?
- even if the answer is yes, please add the caught exception `e` to the new `ProducerClosedException` causes. (There should be a method for this, or maybe it's in the constructor)

Reply:
That's the big question. My idea was that eventually there are only two cases:
- the transaction failed fatally and the consumer had to be closed. Nothing else can be done
- the transaction failed but can be tried again
I wanted to encapsulate these two cases nicely but that's at the expense of hiding the actual exceptions.

Still not sure which is best.

Maybe we can have an extension function like KafkaException.isProducerClosed() ??

Tell me what you think

Reply:
Idk I think the more transparency the better - aka let's just rethrow the existing exceptions. This will mean users won't have to 're-learn' our exceptions, and will be able to rely on the ones they know well already.

I would even argue that I don't see why we should be calling `close()` - it has little to do with transactions themselves, but with handling exceptions. We should probably leave that to the user? 

It looks to me like the user is going to have to wrap us in a try {} catch either way because we are throwing too, so what is the point of writing boilerplate on our end and hiding kafka's exceptions?